Add Physical replication overview pages (#17989)
cockroachdb/docs: src/current/v23.2/physical-cluster-replication-technical-overview.md

---
title: Physical Cluster Replication Technical Overview
summary: Understand the technical details of CockroachDB physical cluster replication.
toc: true
docs_area: manage
---

{{site.data.alerts.callout_info}}
{% include feature-phases/preview.md %}
{{site.data.alerts.end}}

> {% include_cached new-in.html version="v23.2" %} Physical cluster replication uses two clusters with _cluster virtualization_ enabled to stream data from an active primary cluster to a passive standby cluster. Each cluster contains: a _system interface_ and an application _virtual cluster_: {% comment %}Link to Matt's work here on general cluster virtualization reference{% endcomment %}

Reviewer: ```suggestion
{% include_cached new-in.html version="v23.2" %} Physical cluster automatically and continuously streams data from an active _primary_ CockroachDB cluster to a passive _standby_ cluster. Each cluster contains: a _system interface_ and an application _virtual cluster_: {% comment %}Link to Matt's work here on general cluster virtualization reference{% endcomment %}
```
WDYT? This loses a tiny bit of resolution but I think it is easier to parse. "two clusters with CV enabled" seems like it is covered in the other topics.

Author: I like it, I have adapted to this!

{% include {{ page.version.version }}/physical-replication/interface-virtual-cluster.md %}

This separation of concerns means that the replication stream can operate without affecting work happening in the virtual cluster.

> ### Start the replication stream

Reviewer: These sections seem more like "What happens when you...." descriptions rather than tasks, but the section titles seem task-y to me. What if this was called "Replication stream start-up sequence" or something like that?

Author: I updated the first heading as this seemed the most task-y... I think the other two fit in with this update.

> When you initiate physical cluster replication, it will replicate the application virtual cluster from the primary to the standby cluster. The replication stream sets up a series of jobs that complete the following work:

Reviewer 2: Maybe "Starting a physical replication stream consists of two jobs: one on the primary and one on the standby. The job on the standby cluster communicates with the primary cluster via an ordinary SQL connection and is responsible for initiating the replication stream. The stream initialisation proceeds as follows:"

Author: Nice, done!

> Starting a physical replication stream consists of two jobs: one on the primary and one on the standby. The job on the standby cluster communicates with the primary cluster via an ordinary SQL connection and is responsible for initiating the replication stream. The stream initialization proceeds as follows:

Reviewer: What if you named the jobs here? "The physical stream producer job runs on the primary cluster, and the ??? job runs on the standby cluster." And then describe each job? I don't think you mention what the standby cluster's job is.

Author: Added, I think I captured what you were looking for here.

> 1. The standby's system interface communicates to the primary cluster that a stream should start.

Reviewer 2: I would maybe start with:

The standby's system interface connects to the primary cluster and starts the primary cluster's physical stream producer job.

Author: Done!

> 1. The primary cluster chooses a timestamp at which to start the stream and lays down a [protected timestamp]({% link {{ page.version.version }}/architecture/storage-layer.md %}#protected-timestamps), which will protect the entirety of the cluster's data from [garbage collection]({% link {{ page.version.version }}/architecture/storage-layer.md %}#garbage-collection) until an initial scan is complete.

Reviewer 2: I might say "The primary cluster chooses a timestamp at which to start the physical replication stream. Data on the primary is protected from garbage collection until it is replicated to the standby using a protected timestamp."

Author: Done!

> 1. The initial scan runs on the primary and backfills all data from the primary virtual cluster up to the starting timestamp of the replication stream.

Reviewer 2: Maybe "as of the starting timestamp" rather than "up to the starting timestamp"

Author: Done!

1. The primary cluster returns the timestamp and a [job ID]({% link {{ page.version.version }}/show-jobs.md %}#response) for the replication job.

> 1. The standby cluster uses a PG connection to retrieve all of the primary cluster's node addresses. This generates a topology of the primary cluster. If there is a mismatch in nodes between the primary and standby, the replication job will share out the work between the standby's nodes.

Reviewer 2: This and the preceeding step should go above the step "Initial scan runs on the primary" step. Assuming we've mentioned the SQL connection above, I would say:

1. The standby cluster retrieves a list of all nodes in the primary cluster. It uses this list to distribute work across all nodes in the standby cluster.

Author: Done!

> 1. The stream from the primary virtual cluster begins with the standby's system interface ingesting the data into its own virtual cluster.

Reviewer 2: Maybe "Once the initial scan is complete, the primary then begins streaming all changes from the point of the starting timestamp."

Author: Done!

<img src="{{ 'images/v23.2/physical-rep-to.png' | relative_url }}" alt="Two virtualized clusters with system and application tenants showing the directional stream." style="border:0px solid #eee;max-width:100%" />

### During the replication stream

> The replication happens at the byte level, which means that the job is unaware of databases, tables, row boundaries, and so on. However, when a [cutover](#cutover-and-promotion-process) to the standby cluster is initiated, the replication job ensures that the cluster is in a transactionally consistent state as of a certain point in time. Beyond the application data, the job will also replicate users, privileges, basic zone configuration, and schema changes.

Reviewer: ```suggestion
The replication happens at the byte level, which means that the job is unaware of databases, tables, row boundaries, and so on. However, when a [cutover](#cutover-and-promotion-process) to the standby cluster is initiated, the replication job ensures that the cluster is in a transactionally-consistent state as of a certain point in time. Beyond the application data, the job will also replicate users, privileges, basic zone configuration, and schema changes.
```
I wonder if it is worth searching for more instances of this across the various PRs.

Author: We are discussing this, happy to come back to it.

> {% comment %} Add:  no need to coordinate with any schema changes. Don't know anything about zone configurations will be stored using the default zone config.-->{% endcomment %}

Reviewer 2: As of 23.1 we will have basic zone configuration replication as part of the stream.

Author: I have added a sentence here along with some of the other details I hadn't included re what is also replicated. I may have missed something important from this list though?

During the job, [rangefeeds]({% link {{ page.version.version }}/create-and-configure-changefeeds.md %}#enable-rangefeeds) are periodically emitting resolved timestamps (the time where the ingested data is known to be consistent). That is, a guarantee that there are no new writes from before that timestamp. This allows the standby cluster to move the [protected timestamp]({% link {{ page.version.version }}/architecture/storage-layer.md %}#protected-timestamps) forward as the replicated timestamp advances. This information is sent to the primary cluster, which allows for [garbage collection]({% link {{ page.version.version }}/architecture/storage-layer.md %}#garbage-collection) to continue as the replication stream on the standby cluster advances.

{{site.data.alerts.callout_info}}

> If the primary cluster does not receive replicated time information from the standby after 3 days, it cancels the replication job. This ensures that an inactive replication job will not prevent garbage collection. The time at which the job is removed is configurable via the `stream_replication.job_liveness_timeout` [cluster setting]({% link {{ page.version.version }}/cluster-settings.md %}).

Reviewer: Nit: I think we prefer to write out words for single-digit numerals.

Author: I agree, though I think before times and dates you would want to keep as a numeral even for the single-digit numbers.

---
title: Physical Cluster Replication Technical Overview
summary: Understand the technical details of CockroachDB physical cluster replication.
toc: true
docs_area: manage
---

{{site.data.alerts.callout_info}}
{% include feature-phases/preview.md %}
{{site.data.alerts.end}}

{% include_cached new-in.html version="v23.2" %} Physical cluster replication automatically and continuously streams data from an active _primary_ CockroachDB cluster to a passive _standby_ cluster. Each cluster contains: a _system interface_ and an application _virtual cluster_:

{% include {{ page.version.version }}/physical-replication/interface-virtual-cluster.md %}

This separation of concerns means that the replication stream can operate without affecting work happening in the virtual cluster.

### Replication stream start-up sequence

> Starting a physical replication stream consists of two jobs: one on the primary and one on the standby.

Reviewer: ```suggestion
Starting a physical replication stream consists of two jobs: one each on the standby and primary cluster:
```
Adjusted the order to match the list order

Author: Much better, thanks!

- Standby consumer job: Communicates with the primary cluster via an ordinary SQL connection and is responsible for initiating the replication stream. The consumer job ingests updates from the primary cluster producer job.
- Primary producer job: Protects data on the primary cluster and sends updates to the standby cluster.

The stream initialization proceeds as follows:

1. The standby's consumer job connects via its system interface to the primary cluster and starts the primary cluster's physical stream producer job.
1. The primary cluster chooses a timestamp at which to start the physical replication stream. Data on the primary is protected from [garbage collection]({% link {{ page.version.version }}/architecture/storage-layer.md %}#garbage-collection) until it is replicated to the standby using a [protected timestamp]({% link {{ page.version.version }}/architecture/storage-layer.md %}#protected-timestamps).
1. The primary cluster returns the timestamp and a [job ID]({% link {{ page.version.version }}/show-jobs.md %}#response) for the replication job.

> 1. The standby cluster retrieves a list of all nodes in the primary cluster. It uses this list to distribute work across all nodes in the standby cluster.

Reviewer: Does it consider only active / available nodes in the primary cluster? What if there are decommissioning nodes or nodes that just happen to be down when replication is initiated?

Author: So, I checked in with eng. Yes, "it considers nodes it thinks are alive at the point of planning. If it turns out one of those nodes isn't alive or it dies that will result in an error, we will retry and get a new list of nodes".... I have left this as it is for now since a retry is in place. I think that if we get questions around this we may want to expand the detailed section at the bottom of the page rather than here. (I'm also going to open an issue for some follow-up and I will add this as a note to consider.)

> 1. The initial scan runs on the primary and backfills all data from the primary virtual cluster as of the starting timestamp of the replication stream.

Reviewer: Do you intend to limit this to the primary "virtual cluster"? I think replication occurs across the storage cluster, system interface, and application virtual cluster.

Author: huh, the actual unit of replication is a virtual cluster... meaning not the system interface. Or, do you mean the replication job is happening across the whole primary cluster?

(Will add to follow-up)

Reviewer: I am pretty sure that when you cut over to the standby, it has the exact same storage cluster, system interface, and virtual cluster contents as the original primary had, the virtual cluster just has a different name than `application`. I think that's what I meant. I don't think you can cut over the virtual cluster without also getting the primary's storage cluster and system interface configuration. Right? I wouldn't be surprised if I have this wrong.

1. Once the initial scan is complete, the primary then begins streaming all changes from the point of the starting timestamp.

<img src="{{ 'images/v23.2/physical-rep-to.png' | relative_url }}" alt="Two virtualized clusters with system and application tenants showing the directional stream." style="border:0px solid #eee;max-width:100%" />

### During the replication stream

The replication happens at the byte level, which means that the job is unaware of databases, tables, row boundaries, and so on. However, when a [cutover](#cutover-and-promotion-process) to the standby cluster is initiated, the replication job ensures that the cluster is in a transactionally consistent state as of a certain point in time. Beyond the application data, the job will also replicate users, privileges, basic zone configuration, and schema changes.

{% comment %} Add:  no need to coordinate with any schema changes. Don't know anything about zone configurations will be stored using the default zone config.-->{% endcomment %}

> During the job, [rangefeeds]({% link {{ page.version.version }}/create-and-configure-changefeeds.md %}#enable-rangefeeds) are periodically emitting resolved timestamps (the time where the ingested data is known to be consistent). That is, a guarantee that there are no new writes from before that timestamp. This allows the standby cluster to move the [protected timestamp]({% link {{ page.version.version }}/architecture/storage-layer.md %}#protected-timestamps) forward as the replicated timestamp advances. This information is sent to the primary cluster, which allows for [garbage collection]({% link {{ page.version.version }}/architecture/storage-layer.md %}#garbage-collection) to continue as the replication stream on the standby cluster advances.

Reviewer: The first two sentences here are really complex when taken together. I wonder if it can be simplified to avoid the parenthetical and then the "That is.." further explanation.

Author: I have tried to simplify the sentences a little so there aren't too many intros complicating things. Happy to revise again though

{{site.data.alerts.callout_info}}

> If the primary cluster does not receive replicated time information from the standby after 3 days, it cancels the replication job. This ensures that an inactive replication job will not prevent garbage collection. The time at which the job is removed is configurable via the `stream_replication.job_liveness_timeout` [cluster setting]({% link {{ page.version.version }}/cluster-settings.md %}).

Reviewer: Is "replicated time information" the same as or different from "resolved timestamps" in the previous para?

Author: The sends a "heartbeat" to the primary at certain intervals and that heartbeat contains the replicated time. I haven't changed this for now as I feel like we would then have to spell out what info is in the heartbeat, and I think that might be getting too deep. Again, another place to evaluate if customers start to have questions around this kind of technical detail. Thanks for bringing this up
